### PR TITLE
Jenkins api

### PR DESCRIPTION
**Pull request by emizell**
Please review changes made to support the Jenkins API

### CODE REVIEW

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	
+	//check if event matches the pattern - returns true if matches
+	private static boolean evaluateEvent(EventResult event, Pattern pattern) {
+		
+		if (pattern == null ) {
+			return true;
+		}
+		
+		String json = new Gson().toJson(event);
+		boolean result = !pattern.matcher(json).find();
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Is this `!` on purpose?
The comment above the method seems to indicate we should return true for a match, unless I'm missing something?

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	
+	//filter in or out events based on the pattern 
+	private static Collection<EventResult> filterEvents(Collection<EventResult> events, String regexFilter) {
+		if (regexFilter == null) {
+			return events;
+		}
+		
+		List<EventResult> returnEvents = new ArrayList<EventResult>();	
+		for (Iterator iterator = events.iterator(); iterator.hasNext();) {
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Any reason this isn't a `for (EventResult event : events)` ?
I don't see `iterator` being used inside the loop (for `remove`, for instance)

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			EventResult event = (EventResult) iterator.next();
+			if (evaluateEvent(event, getPattern(regexFilter))) {
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Probably not too significant, but as `getPattern` internally calls `Pattern.compile`, it's better to extract the result `Pattern` object to a local variable outside of this loop, to avoid some overhead.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			result.add(newEvent);
+		}
+		return result;
+	}
+	
+	//get the arclink for the event
+	public static String getArcLink(ApiClient apiClient, String eventId,
+			RegressionInput input, DateTime activeWindowStart) {
+
+		DateTime activeWndowStart = activeWindowStart;
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Seems like `activeWndowStart` isn't needed when we got `activeWindowStart` :)

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			result.add(newEvent);
+		}
+		return result;
+	}
+	
+	//find new errors from error list
+	private static List<OOReportEvent> getNewErrors (ApiClient apiClient, RegressionInput input, Collection<EventResult> events, DateTime deploymentStart) {		
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();		
+		for (EventResult event : events) {			
+			if (!input.deployments.contains(event.introduced_by)) {
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Probably better to use `!CollectionUtil.safeContains(input.deployments, event.introduced_by)`, as it protects us in case `input.deployments` is `null`

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+		}
+		return result;
+	}
+	
+	//find resurfaced errors from error list
+	private static List<OOReportEvent> getResurfacedErrors(ApiClient apiClient, RegressionInput input,
+			Collection<EventResult> events, DateTime deploymentStart) {
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();
+		for (EventResult event : events) {
+			if (!event.labels.contains(RegressionStringUtil.RESURFACED_ISSUE)) {
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Probably better to use `!CollectionUtil.safeContains(event.labels, RegressionStringUtil.RESURFACED_ISSUE)`, as it protects us in case `input.deployments` is `null`

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+			Collection<EventResult> events, int topIssuesVolume, DateTime deploymentStart) {
+		List<OOReportEvent> result = new ArrayList<OOReportEvent>();
+		
+		int counter = 1;
+		for (EventResult event : events) {
+			String arcLink = getArcLink(apiClient, event.id, input, deploymentStart);
+			OOReportEvent newEvent = new OOReportEvent(event, null, arcLink);
+			result.add(newEvent);
+			
+			if (counter == topIssuesVolume) {
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Can check `topIssuesVolume` directly against `result.size()`
(The `size` method for `ArrayList` runs at constant time, and is optimized by the JVM so no performance overhead for calling it a lot)

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RateRegression.java [hunk 8]
```diff
@@ -54,15 +54,15 @@ public int compare(RegressionResult o1, RegressionResult o2) {
...
+	public static List<EventResult> getSortedNewEvents(Collection<EventResult> events) {
 
 		List<EventResult> result = Lists.newArrayList(events);
 
 		result.sort(new Comparator<EventResult>() {
 
 			@Override
 			public int compare(EventResult o1, EventResult o2) {
-				return (int) (o1.stats.hits - o2.stats.hits);
+				return (int) (o2.stats.hits - o1.stats.hits);
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/regression/RateRegression.java]: Note this method is used in returning values all over this class, so changing the sort direction affects other places as well.

Is this change intended to change all of them as well?
If not, it should be reverted, and a new method (or overload with a boolean param for direction) should be used

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java [hunk 9]
```diff
@@ -23,8 +23,10 @@
 	public String viewId;
 
 	public DateTime activeWindowStart;
+	public DateTime nonRegressionActiveWindowStart;
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java]: I don't see where this is used.
Is it used someplace outside of the `api-client`?

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java [hunk 10]
```diff
@@ -23,8 +23,10 @@
...
 	public int activeTimespan;
 	public int baselineTimespan;
+	public String baselineTime;
```
Comment (dor-levi) [inline api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java]: I don't see where this is used.
Is it used someplace outside of the `api-client`?

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 1 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: We are looking for matches in the pattern. The ! would return a false if it doesn't match. I guess this is a negative test.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 2 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Good call. I copied this code and never updated the iterator. We are not removing from the list, just returning a new list with what has been filtered.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 3 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Again, copied code. Happy to change if needed.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 4 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Funny, didn't event notice. Copied from API client. Changed

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 5 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: There is a check before this code is called that makes sure there are events. No need to check again.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 6 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: There is a check before this code is called that makes sure there are events. No need to check again.

File: api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java (hunk 7 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/cicd/ProcessQualityGates.java]: Nice. Changed

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RateRegression.java (hunk 8 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/regression/RateRegression.java]: This is interesting. I found this code repeated 5 or so times in the API client. I am a fan of reusable code. The sort should bubble up the items with the most hits. I think is was a bug, but we should check

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java (hunk 9 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java]: Lingering garbage I forgot to remove.

File: api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java (hunk 10 above)
Comment (emizell) [reply api-client-util/src/main/java/com/takipi/api/client/util/regression/RegressionInput.java]: Lingering garbage I forgot to remove

**Comment by emizell** [comment]
I pushed the changes to the api-client
